test/common/config/grpc_stream_test.cc

```diff
@@ -0,0 +1,134 @@
+                     random_, dispatcher_, stats_, rate_limit_settings_,
+                     [this](std::unique_ptr<envoy::api::v2::DiscoveryResponse>&& message) {
+                       received_message_ = *message;
+                     }) {}
+
+  //  std::string version_;
+  //  const Protobuf::MethodDescriptor* method_descriptor_;
+  //  NiceMock<Upstream::MockClusterManager> cm_;
+  NiceMock<Event::MockDispatcher> dispatcher_;
+  //  Event::MockTimer* timer_;
+  //  Event::TimerCb timer_cb_;
+  //  envoy::api::v2::core::Node node_;
+  //  NiceMock<Config::MockSubscriptionCallbacks<envoy::api::v2::ClusterLoadAssignment>> callbacks_;
+  Grpc::MockAsyncStream async_stream_;
+  //  std::string last_response_nonce_;
+  //  std::vector<std::string> last_cluster_names_;
+  //  NiceMock<LocalInfo::MockLocalInfo> local_info_;
+  //  Event::MockTimer* init_timeout_timer_;
+  Stats::IsolatedStoreImpl stats_;
+  NiceMock<Runtime::MockRandomGenerator> random_;
+  Envoy::Config::RateLimitSettings rate_limit_settings_;
+  NiceMock<MockXdsGrpcContext> xds_grpc_context_;
+  std::unique_ptr<Grpc::MockAsyncClient> async_client_owner_;
+  Grpc::MockAsyncClient* async_client_;
+  envoy::api::v2::DiscoveryResponse received_message_;
+
+  GrpcStream<envoy::api::v2::DiscoveryRequest, envoy::api::v2::DiscoveryResponse> grpc_stream_;
+};
+
+// Tests that establishNewStream() establishes it, a second call does nothing, and a third call
+// after the stream was disconnected re-establishes it.
+TEST_F(GrpcStreamTest, EstablishNewStream) {
+  EXPECT_FALSE(grpc_stream_.grpcStreamAvailable());
+  // Successful establishment
+  {
+    EXPECT_CALL(*async_client_, start(_, _)).WillOnce(Return(&async_stream_));
+    EXPECT_CALL(xds_grpc_context_, handleStreamEstablished());
+    grpc_stream_.establishNewStream();
+    EXPECT_TRUE(grpc_stream_.grpcStreamAvailable());
+  }
+  // Idempotency: do nothing (other than logging a warning) if already connected
+  {
+    EXPECT_CALL(*async_client_, start(_, _)).Times(0);
+    EXPECT_CALL(xds_grpc_context_, handleStreamEstablished()).Times(0);
+    grpc_stream_.establishNewStream();
+    EXPECT_TRUE(grpc_stream_.grpcStreamAvailable());
+  }
+  grpc_stream_.onRemoteClose(Grpc::Status::GrpcStatus::Ok, "");
+  EXPECT_FALSE(grpc_stream_.grpcStreamAvailable());
+  // Successful re-establishment
+  {
+    EXPECT_CALL(*async_client_, start(_, _)).WillOnce(Return(&async_stream_));
+    EXPECT_CALL(xds_grpc_context_, handleStreamEstablished());
+    grpc_stream_.establishNewStream();
+    EXPECT_TRUE(grpc_stream_.grpcStreamAvailable());
+  }
+}
+
+TEST_F(GrpcStreamTest, FailToEstablishNewStream) {
+  EXPECT_CALL(*async_client_, start(_, _)).WillOnce(Return(nullptr));
+  EXPECT_CALL(xds_grpc_context_, handleEstablishmentFailure());
+  grpc_stream_.establishNewStream();
+  EXPECT_FALSE(grpc_stream_.grpcStreamAvailable());
+}
+
+TEST_F(GrpcStreamTest, SendMessage) {
```

Comment:
Please add `//` above each with a TL;DR of what the test does.

Reply:
Done.